### DISCUSSION TITLE
`TypeError` for generic `dict` parameter when `pydantic=True` in Celery 5.5

### DISCUSSION THREAD

**Discussion by mmangus** [celery/celery#9316] [Q&A]
I am excited to see that Celery 5.5 will add support for Pydantic models. This was a major selling point for me, because I'm adding Celery to a project that already makes extensive use of Pydantic.

When testing this functionality, I have found that setting `pydantic=True` in my task decorator throws a `TypeError` for this task, apparently due to the annotation of the `max_urls` parameter:

```python
@app.task(pydantic=True)
def parse_urls(
    body: str,
    message_id: UUID,
    max_urls: dict[str, int],
) -> SomePydanticModel: ...
```

```python
celery-pool-1     | Traceback (most recent call last):
celery-pool-1     |   File "/usr/local/lib/python3.12/site-packages/celery/app/trace.py", line 453, in trace_task
celery-pool-1     |     R = retval = fun(*args, **kwargs)
celery-pool-1     |                  ^^^^^^^^^^^^^^^^^^^^
celery-pool-1     |   File "/usr/local/lib/python3.12/site-packages/celery/app/trace.py", line 736, in __protected_call__
celery-pool-1     |     return self.run(*args, **kwargs)
celery-pool-1     |            ^^^^^^^^^^^^^^^^^^^^^^^^^
celery-pool-1     |   File "/usr/local/lib/python3.12/site-packages/celery/app/base.py", line 132, in wrapper
celery-pool-1     |     if issubclass(arg_annotation, BaseModel):
celery-pool-1     |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
celery-pool-1     |   File "<frozen abc>", line 123, in __subclasscheck__
celery-pool-1     | TypeError: issubclass() arg 1 must be a class
```

As shown in the traceback, it is throwing from this check:
https://github.com/celery/celery/blob/ad882dc991e46b46c1a195771a897de7127f1e6b/celery/app/base.py#L132

This check cannot work if any annotation is not a class, as indicated by the error thrown. In my case, some manual inspection in an interactive shell confirms that the problematic annotation is the `max_urls` dictionary:

```python
>>> issubclass(inspect.signature(parse_urls).parameters['max_urls'].annotation, BaseModel)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[9], line 1
----> 1 issubclass(inspect.signature(parse_urls).parameters['max_urls'].annotation, BaseModel)

File <frozen abc>:123, in __subclasscheck__(cls, subclass)

TypeError: issubclass() arg 1 must be a class
```

I believe the root cause here is that parameterized generic forms of builtin types like `dict[str, str]` or `list[int]` are not classes.

```python
>>> from inspect import isclass
>>> isclass(dict)
True

>>> isclass(dict[str, str])
False

>>> isclass(list[int])
False
```

Q: **Is my use case considered valid? Or is it intended that any task that opts into Pydantic serialization must exclusively use Pydantic models for all parameters?**

I would be happy to contribute a fix to Celery here if appropriate. It seems to me that the simplest fix for this problem would be to add an extra guard on `app/base.py:132`, like:
```python3
from inspect import isclass
...
        if isclass(arg_annotation) and issubclass(arg_annotation, BaseModel):
```

**Comment by Nusnus** [comment]
@mathiasertl

**Comment by mathiasertl** [comment]
I have a fix.

@Nusnus, is the project committed to still supporting Python 3.8 in the next release? End of life is literally on Tuesday:

https://devguide.python.org/versions/#versions

Reason I'm asking is that supporting Python3.8 is more complicated here.

**Reply by Nusnus** [reply]
Yes.
That being said, we can maybe allow this feature to not be fully supposed in 3.8 to save you the trouble that will be irrelevant soon anyways.

So you have my green light on skipping that part 👍
